File: src/main/scala/Bowling.scala
trait BowlingFrame {
  def scores: List[Int]
  def isComplete: Boolean
}

case class NotLastFrame(scores: List[Int]) extends BowlingFrame {
  def isStrike: Boolean = scores.length == 1 && scores.head == 10
  def isSpare: Boolean = scores.length == 2 && scores.sum == 10

  def isComplete: Boolean = isStrike || scores.length == 2
}

case class LastFrame(scores: List[Int]) extends BowlingFrame {
  def isComplete: Boolean = (scores.length == 2 && scores.sum < 10) || scores.length == 3
}

object Bowling {

  private def getNumericalValueFromScoreAndLastFrame(lastOptionFrame: Option[BowlingFrame], score: String): Int = {
    (lastOptionFrame, score) match {
      case (_, "X") => 10
      case (Some(frame), "/") => 10 - frame.scores.last
      case (_, intValue) => intValue.toInt
    }
  }

  def frames(individualScores: List[String]): List[BowlingFrame] = {
    individualScores.foldLeft(List.empty[BowlingFrame])((acc, d) => {
      val numericalValue: Int = getNumericalValueFromScoreAndLastFrame(acc.lastOption, d)
      (acc.lastOption, acc.length) match {
        case (Some(notLastFrame: NotLastFrame), accLength) =>
          if (notLastFrame.isComplete) {
            val newFrame = if (accLength == 9) LastFrame(List(numericalValue)) else NotLastFrame(List(numericalValue))
            acc :+ newFrame
          } else {
            acc.init ++ List(NotLastFrame(notLastFrame.scores :+ numericalValue))
          }
        case (Some(lastFrame: LastFrame), 10) => acc.init ++ List(LastFrame(lastFrame.scores :+ numericalValue))
        case (None, 0) => List(NotLastFrame(List(numericalValue)))
      }
    })
  }

  private def getScoreForCompleteNotLastFrame(notLastFrame: NotLastFrame, frameNumber: Int, frames: List[BowlingFrame]): Option[Int] = {
    val sumOfFrameScores = notLastFrame.scores.sum
    (notLastFrame.isStrike, notLastFrame.isSpare) match {
      case (false, false) => Option(sumOfFrameScores)
//      TODO: IMPLEMENT BELOW
      case (true, false) => None
      case (false, true) => None
    }
  }

  private def getScoresFromFrame: List[BowlingFrame] => (BowlingFrame, Int) => Option[Int] = {
    frames => (bowlingFrame, frameNumber) => {
      (bowlingFrame, bowlingFrame.isComplete) match {
        case (_, false) => None
        case (_: LastFrame, true) => Option(bowlingFrame.scores.sum)
        case (notLastFrame: NotLastFrame, true) => getScoreForCompleteNotLastFrame(notLastFrame, frameNumber, frames)
      }
    }
  }

  def scores(frames: List[BowlingFrame]): List[Option[Int]] = {
    val curriedScoresFromFrame: (BowlingFrame, Int) => Option[Int] = getScoresFromFrame(frames)
    val individualFrameScores: List[Option[Int]] = frames.zipWithIndex.map(d => curriedScoresFromFrame(d._1, d._2))
//    Can below use reduceLeft?
    individualFrameScores.foldLeft(List.empty[Option[Int]])((acc, d) => {
      if (acc.isEmpty) List(d) else {
        val lastFrameScore = for {
          totalLastScore <- acc.last
          individualFrameScore <- d
        } yield totalLastScore + individualFrameScore
        acc :+ lastFrameScore
      }
    })
  }
}

File: js/bowling.js
const sum = numbers => numbers.reduce((acc, d) => acc + d, 0);

const getNumericalValueFromScoreAndLastFrame = (lastFrame, score) => {
    if (score == "X") return 10;
    if (score == "/") return 10 - lastFrame.scores[lastFrame.scores.length - 1];
    return parseInt(score, 10);
}

const NotLastFrame = scores => {
    const isStrike = scores.length == 1 && scores[0] == 10
    return {
        scores,
        isStrike,
        isSpare: scores.length == 2 && sum(scores) == 10,
        isComplete: isStrike || scores.length == 2,
        isNotLastFrame: true
    }
};

const LastFrame = scores => {
    return {
        scores,
        isComplete: (scores.length == 2 && sum(scores) < 10) || scores.length == 3,
        isNotLastFrame: false
    }
};

function frames(individualScores) {
    return individualScores.reduce((acc, d) => {
        const accLength = acc.length;
        const lastFrameInAcc = acc.length ? acc[acc.length - 1] : null;
        const numericalValue = getNumericalValueFromScoreAndLastFrame(lastFrameInAcc, d);
        if (!lastFrameInAcc) return [NotLastFrame([numericalValue])];
        if (lastFrameInAcc.isComplete) {
            const newFrame = accLength == 9 ? LastFrame([numericalValue]) : NotLastFrame([numericalValue]);
            return acc.concat([newFrame]);
        } else {
            const newScores = lastFrameInAcc.scores.concat([numericalValue]);
            const newFrame = lastFrameInAcc.isNotLastFrame ? NotLastFrame(newScores) : LastFrame(newScores);
            return acc.slice(0, -1).concat([newFrame]);
        }

    }, []);
}

const getScoreForCompleteNotLastFrame = (notLastFrame, subsequentFrames) => {
    const sumOfFrameScores = sum(notLastFrame.scores);
    const subsequentScores = subsequentFrames.flatMap(frame => frame.scores);
    if (notLastFrame.isStrike) {
        return subsequentScores.length >= 2 ? sumOfFrameScores + subsequentScores[0] + subsequentScores[1] : null;
    } else if (notLastFrame.isSpare) {
        return subsequentScores.length >= 1 ? sumOfFrameScores + subsequentScores[0] : null;
    } else {
        return sumOfFrameScores;
    }
}

const getScoresFromFrame = frames => (bowlingFrame, frameNumber) => {
    if (!bowlingFrame.isComplete) return null;
    if (!bowlingFrame.isNotLastFrame) return sum(bowlingFrame.scores)
    return getScoreForCompleteNotLastFrame(bowlingFrame, frames.slice(frameNumber + 1));
}

function scores(frames) {
    const curriedScoresFromFrame = getScoresFromFrame(frames)
    const individualFramesScores = frames.map((frame, index) => curriedScoresFromFrame(frame, index));
    return individualFramesScores.reduce((acc, d) => {
        if (acc.length == 0) {
            return [d];
        } else {
            const totalLastScore = acc[acc.length - 1];
            const nextScore = (totalLastScore && d) ? totalLastScore + d : null; 
            return acc.concat([nextScore]);
        }
    }, [])
}

const Bowling = {
    frames,
    scores
};

exports.NotLastFrame = NotLastFrame;
exports.LastFrame = LastFrame;
exports.Bowling = Bowling;